### ISSUE TITLE
SSH tunnel on iOS doesn’t work

### ISSUE THREAD

**Issue by alexs279** [isontheline/pro.webssh.net#584]
# Bug description
Setup ssh tunnel to Amazon ec2 for port 8080:localhost:8080.  After logging in safari just hangs trying to access. 
I have tried same configuration  with another app called remoter vnc and it worked fine. 

# Screenshots
_If applicable, add screenshots to help explain your problem_

# Steps to reproduce
1. On iOS device Setup ssh with pem key to Amazon. 
2. Setup a port redirect  to localhost
3. Use safari or chrome to try and access localhost:port
4. Browser freezes trying to connect

**Comment by isontheline** [comment]
Hello @alexs279 👋

Could you try these tips ?
* Run WebSSH in split view
OR
* Launch your tunnel by using "VPN-Over-SSH" option

**Comment by alexs279** [comment]
Hi isonline!

I just tried vpn over ssh option and same problem.  
I am on my phone so can’t use splitview but when I go back and forth between browser and  WebSSH app I still see the connection active and I have to manually press it to disconnect.

**Comment by isontheline** [comment]
I've just tested and my phone and seem to works.

Could you screenshot ?
Only the port forwarding form field.

Thanks

**Comment by alexs279** [comment]
![32ACCBD3-0453-423E-BDD3-F21CD7BAEF52](https://user-images.githubusercontent.com/95941410/161269975-27758bd6-aaab-4d25-a4ed-9865aa9a3d21.jpeg)

**Comment by isontheline** [comment]
Strange 😬

When the tunnel is active, could you do that :
WebSSH > mashREPL > curl -vv http://localhost:8080

**Comment by alexs279** [comment]
![4C5CF284-2CF7-49CC-BD92-35182A7E069A](https://user-images.githubusercontent.com/95941410/161271018-80f53991-5742-491f-a1e0-5fa46fcbeed5.png)


Btw I setup the same ssh connection and able to log in fine.  So it’s really just the tunnel that doesn’t work.

**Comment by isontheline** [comment]
Screenshot tou sent (mashREPL) shows that tunnel works!

403 Forbidden

Do you agree?

**Comment by isontheline** [comment]
Please share your other app screenshot settings.

It seems your webserver should allow 127.0.0.1 connections.

**Comment by alexs279** [comment]
![E1B15CBB-89AD-415C-8615-2C21C9BDB170](https://user-images.githubusercontent.com/95941410/161272860-38d46629-8988-48f3-b491-630344bb5faa.png)


Maybe the redirect not picking up the right port from the app.   Meaning is trying not 8080 but something else like 80

**Comment by isontheline** [comment]
Thanks for sharing!

Could you check which IP subnets are allowed inside your Jenkins server?

**Comment by alexs279** [comment]
the eth0 is at /20 or 255.255.240.0 

no other interfaces

**Comment by isontheline** [comment]
There is an apache / httpd in front ?

**Comment by alexs279** [comment]
Its a jenkins server so I believe they do use tomcat/apache

**Comment by alexs279** [comment]
Here is more information.

On the ec2 server I ran this and tried to redirect port 8080:localhost:80. and this is the message on ec2 webserver

sudo python3 -m http.server 80

Serving HTTP on 0.0.0.0 port 80 (http://0.0.0.0:80/) ...
127.0.0.1 - - [01/Apr/2022 14:35:09] "GET / HTTP/1.1" 200 -
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 45078)
Traceback (most recent call last):
  File "/usr/lib64/python3.7/socketserver.py", line 650, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib64/python3.7/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib64/python3.7/http/server.py", line 646, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/lib64/python3.7/socketserver.py", line 720, in __init__
    self.handle()
  File "/usr/lib64/python3.7/http/server.py", line 426, in handle
    self.handle_one_request()
  File "/usr/lib64/python3.7/http/server.py", line 414, in handle_one_request
    method()
  File "/usr/lib64/python3.7/http/server.py", line 653, in do_GET
    self.copyfile(f, self.wfile)
  File "/usr/lib64/python3.7/http/server.py", line 844, in copyfile
    shutil.copyfileobj(source, outputfile)
  File "/usr/lib64/python3.7/shutil.py", line 82, in copyfileobj
    fdst.write(buf)
  File "/usr/lib64/python3.7/socketserver.py", line 799, in write
    self._sock.sendall(b)
BrokenPipeError: [Errno 32] Broken pipe
----------------------------------------



When I do the same with my other app "remoter vnc" I get the following on the ec2 server
127.0.0.1 - - [01/Apr/2022 14:36:21] "GET / HTTP/1.1" 200 -

**Comment by jarrah31** [comment]
I know it seems daft but what version of WebSSH are you running? I ask because I tried tunnelling and found it didn't work on safari, but then I realised I was on an older WebSSH version. Updating to the latest 17.4 fixed it for me.

**Comment by alexs279** [comment]
I’m on 17.4.887

i just downloaded a couple days ago.

**Comment by isontheline** [comment]
I understand that other apps are working as you expect but the test you have done with mashREPL says that your webserver denies your access. 

I don't know where I could make a fix for this issue as the tunnel is "working as expected" from my point of view.

Couldn't you allow your webserver to accept loopback connections?

**Comment by isontheline** [comment]
I'm able to run with success the same following test as you on a VPS :
<img width="1409" alt="image" src="https://user-images.githubusercontent.com/44212923/161330148-dac31661-f713-4585-b96e-d682fa3e219d.png">

It's strange that you obtain your error

**Comment by alexs279** [comment]
Ok here is something interesting. 

I use the test http at 80

I ran the mashREPL curl command. 

I get back a response with the info from server but then it gives me a connection error. 

![71D71C35-79BD-4667-A45B-04FAE5F22048](https://user-images.githubusercontent.com/95941410/161332104-00265294-4ddd-42e1-9c49-96d44dbeb8bf.png)
![99DCDD03-6CEC-4C97-A07D-4D837A2E353E](https://user-images.githubusercontent.com/95941410/161332108-ba823037-cd13-4779-a241-72a2b15dfd66.png)

**Comment by isontheline** [comment]
Seems localhost is an alias to the IPv6 loopback address

Does `curl -vv http://127.0.0.1` is better?

If you use this forward rule, does it works better? => `8080:127.0.0.1:8080`

It will force the use of IPv4

**Comment by alexs279** [comment]
Same. 

![22B5D4A5-1A45-4E46-B233-50E2C3DA3B60](https://user-images.githubusercontent.com/95941410/161334150-7c006a2f-de3b-4387-852a-25175e788aff.png)

**Comment by alexs279** [comment]
Seems no error but browser still times out. 

![60F51F2A-64A6-4951-AA66-AD3DC4ECA4DC](https://user-images.githubusercontent.com/95941410/161334458-b4cac63d-dec3-4c1d-9b47-371a46940ab4.png)

**Comment by isontheline** [comment]
Are you sure that "VPN-Over-SSH" didn't solve your problem when using rule ```
8080:127.0.0.1:8080
```

**Comment by alexs279** [comment]
That worked!!!

we never tried that option.  I originally tried vpn over ssh when I had 8080:localhost:8080

now with 127.0.0.1 vpn over ssh works 👍

thanks for the help!

**Comment by isontheline** [comment]
You're Welcome!

I will improve WebSSH to fix localhost to IPv4

Have a nice day ☀️

**Comment by melonsode** [comment]
hi, I had the same problem. I was very happy when this app had got muliti port forwarding on Test flight version. I had confirmed multi port forwarding successfully work and I didn’t need to use this app for a while. Recently I tried to use this app with multi port forwarding, then it doesn’t work. Not inly mutiport fowading but also single port forwarding not work well. 
Today, I have found this issue. I made the option ‘VPN-over-SSH’ turned on. Then multiport fowarding works again.

I don’t know what this option means. In older version of this app, I had left this option off but single port forwading had worked.
I suggest some explanation should be added in help page. Thank you.

**Comment by isontheline** [comment]
Hi @melonsode 

Multi Port Forwarding should wok even without VPN-Over-SSH.

This option is only needed when you need to use forwarded ports outside WebSSH : https://webssh.net/documentation/help/networking/vpn-over-ssh/

**Comment by melonsode** [comment]
Thank you for replying. I use this app to use port forwarding with Safari.
Although I don’t remember if I had turned on the “VPN-Over-SSH” option or not 
when multi port forwarding was not included in this app, I was able to use port forward with Safari. Was the option on in default setting?

Anyway, in my case, using port forward out of this app, I should turn on the “VPN-Over-SSH” option. I understand it. Thank you.
